### DISCUSSION TITLE
How to fix a runtime warning?

### DISCUSSION THREAD

**Discussion by decadance-dance** [mosecorg/mosec#571] [General]
I almost always get such message running my CV models: `RuntimeWarning: IPC data (x bytes) is large, which may affect performance`.
What can I do to fix this warning to ensure I am OK and it's not gonna affect performance.
P.S. I use my own mixins:
```python
from typing import Any

import pickle
import numpy as np

from mosec.errors import DecodingError, EncodingError


class PickleMixin:
    """Msgpack worker mixin interface."""

    # pylint: disable=no-self-use

    def serialize(self, data: np.ndarray) -> bytes:
        """Serialize with msgpack for the last stage (egress).

        Arguments:
            data: the **same type** as returned by
                :py:meth:`Worker.forward <mosec.worker.Worker.forward>`

        Returns:
            the bytes you want to put into the response body

        Raises:
            EncodingError: if the data cannot be serialized with msgpack

        """
        
        try:
            data_bytes = pickle.dumps(data, fix_imports=False, protocol=pickle.HIGHEST_PROTOCOL)
        except Exception as err:
            raise EncodingError from err
        return data_bytes  # type: ignore

    def deserialize(self, data: bytearray) -> Any:
        """Deserialize method for the first stage (ingress).

        Arguments:
            data: the raw bytes extracted from the request body

        Returns:
            the **same type** as the input of
            :py:meth:`Worker.forward <mosec.worker.Worker.forward>`

        Raises:
            DecodingError: if the data cannot be deserialized with msgpack

        """
        try:
            data_msg = pickle.loads(data, fix_imports=False)
        except Exception as err:
            raise DecodingError from err
        return data_msg
```

**Comment by kemingy** [comment]
This warning indicates that the data you passed from one worker to the next worker is huge. This will affect the performance since it may hit the memory bandwidth.

To reduce the size:
1. reduce the unnecessary part of the data (depending on your use case)
2. data compression (dense matrix may not be compressed anymore, JPEG format can be a choice if this doesn't affect the model precision)
3. better IPC serialization (`pickle` is good enough, others might reduce the size a little bit)
4. try to do everything in one worker (need to benchmark, this might make it worse)
5. try the [redis IPC mixin](https://mosec.readthedocs.io/en/latest/examples/ipc.html#redis-py) (need to benchmark)

For CV tasks, this might be an impossible task to reduce the size. You can ignore this warning.

**Reply by decadance-dance** [reply]
ok, thank you for such detailed answer
